Implement SAML custom attributes support for Identity Provider

lloydmeta opened the pull request:
This commit adds support for custom attributes in SAML single sign-on requests in the Elasticsearch X-Pack Identity Provider plugin. This feature allows passage of custom key-value attributes in SAML responses from the init endpoint. 

~I opted for an array of `attributes` instead of an object/map because it felt easier and avoid mapping explosion if ingested somewhere, but could go the other way as well.~

Example
```js
POST /_idp/saml/init
// Body
{
  "entity_id": "https://example.com/saml/sp",
  "acs": "https://example.com/saml/acs",
  "authn_state": "some-authn-state-value",
  // Added
  "attributes": { 
    "department": [ "engineering", "product" ],
    "region": [ "APJ" ] }
  }
}
```

Key components:
- Added SamlInitiateSingleSignOnAttributes class for holding attributes
- Added validation for null, and empty attribute keys
- Updated request and response objects to handle attributes
- Added test coverage to validate attributes functionality

**Note** : this PR was generated with the help of Windsurf as part of our internal pilot.

Related to IAM-1789

lloydmeta committed:
Implement SAML custom attributes support for Identity Provider

This commit adds support for custom attributes in SAML single sign-on requests
in the Elasticsearch X-Pack Identity Provider plugin. This feature allows
passage of custom key-value attributes in SAML requests and responses.

Key components:
- Added SamlInitiateSingleSignOnAttributes class for holding attributes
- Added validation for null and empty attribute keys
- Updated request and response objects to handle attributes
- Modified authentication flow to process attributes
- Added test coverage to validate attributes functionality

The implementation follows Elasticsearch patterns with robust validation
and serialization mechanisms, while maintaining backward compatibility.

lloydmeta committed:
Add test for SAML custom attributes in authentication response

This commit adds a comprehensive test that verifies SAML custom attributes
are correctly handled in the authentication response builder. The test ensures:

1. Custom attributes with single and multiple values are properly included
2. The response with custom attributes is still correctly signed
3. The XML schema validation still passes with custom attributes
4. We can locate and verify individual attribute values in the response

This provides critical test coverage for the SAML custom attributes
feature implementation.

lloydmeta commented:
Interesting...[build](https://buildkite.com/elastic/elasticsearch-serverless-es-pr-check/builds/25500/summary/annotations?sid=0196ecdf-a01b-4751-9cad-ff27deb2305d#0196ecdf-a08f-4c7e-984c-acbd4a99808c/89-2660) is failing with something that was untouched

```
Note: Recompile with -Xlint:removal for details.
/dev/shm/bk/bk-agent-prod-gcp-1747730850111949647/elastic/elasticsearch-serverless-es-pr-check/modules/stateless/src/main/java/co/elastic/elasticsearch/stateless/objectstore/ObjectStoreService.java:302: error: cannot find symbol
        for (var projectId : event.projectDelta().added()) {
                                  ^
  symbol:   method projectDelta()
  location: variable event of type ClusterChangedEvent
/dev/shm/bk/bk-agent-prod-gcp-1747730850111949647/elastic/elasticsearch-serverless-es-pr-check/modules/stateless/src/main/java/co/elastic/elasticsearch/stateless/objectstore/ObjectStoreService.java:330: error: cannot find symbol
        for (var projectId : event.projectDelta().removed()) {
                                  ^
  symbol:   method projectDelta()
  location: variable event of type ClusterChangedEvent
2 errors
```

lloydmeta committed:
Add backward compatibility overload for SuccessfulAuthenticationResponseMessageBuilder.build

This commit adds an overloaded build method that accepts only two parameters
(user and authenticationState) and forwards the call to the three-parameter
version with null for the customAttributes parameter. This maintains backward
compatibility with existing code that doesn't use custom attributes.

This fixes a compilation error in ServerlessSsoIT.java which was still using
the two-parameter method signature.

Signed-off-by: lloydmeta <lloydmeta@gmail.com>

lloydmeta commented on x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java:
```diff
@@ -199,6 +229,16 @@ private AttributeStatement buildAttributes(UserServiceAuthentication user) {
             attributes.add(name);
         }
+        // Add custom attributes if provided
+        if (customAttributes != null && customAttributes.getAttributes().isEmpty() == false) {
```
Can't do this; stylechecker will not like it.

lloydmeta committed:
Add validation for duplicate SAML attribute keys

This commit enhances the SAML attributes implementation by adding validation
for duplicate attribute keys. When the same attribute key appears multiple
times in a request, the validation will now fail with a clear error message.

Signed-off-by: lloydmeta <lloydmeta@gmail.com>

lloydmeta committed:
Refactor SAML attributes validation to follow standard patterns

This commit improves the SAML attributes validation by:

1. Adding a dedicated validate() method to SamlInitiateSingleSignOnAttributes
   that centralizes validation logic in one place
2. Moving validation from constructor to dedicated method for better error reporting
3. Checking both for null/empty keys and duplicate keys in the validate() method
4. Updating SamlInitiateSingleSignOnRequest to use the new validation method
5. Adding comprehensive tests for the new validation approach

These changes follow standard Elasticsearch validation patterns, making the
code more maintainable and consistent with the rest of the codebase.

elasticsearchmachine commented:
Hi @lloydmeta, I've created a changelog YAML for you.

lloydmeta committed:
Update docs/changelog/128176.yaml

lloydmeta committed:
Improve SAML response validation in identity provider tests

Enhanced the testCustomAttributesInIdpInitiatedSso test to properly validate
both SAML response structure and custom attributes using DOM parsing and XPath.

Key improvements:
- Validate SAML Response/Assertion elements exist
- Precisely validate custom attributes (department, region) and their values
- Use namespace-aware XML parsing for resilience to format changes

Signed-off-by: lloydmeta <lloydmeta@gmail.com>

tvernum commented on x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/rest/action/RestSamlInitiateSingleSignOnAction.java:
```diff
@@ -41,6 +42,11 @@ public class RestSamlInitiateSingleSignOnAction extends IdpBaseRestHandler {
             (p, c) -> SamlAuthenticationState.fromXContent(p),
             new ParseField("authn_state")
         );
+        PARSER.declareObject(
```
I was trying to work out how/why this works if your JSON looks like the PR description.
It turns out, that it doesn't.

The test sends, and the rest handler expects a body like this:
```
{
  "entity_id": "ec:123456:abcdefg",
  "acs": "https://sp1.test.es.elasticsearch.org/saml/acs",
  "attributes": {
    "attributes": [
      {
        "key": "department",
        "values": [
          "engineering",
          "product"
        ]
      },
      {
        "key": "region",
        "values": [
          "APJ"
        ]
      }
    ]
  }
}
```

I'd prefer we just use a JSON object
```
{
  "entity_id": "ec:123456:abcdefg",
  "acs": "https://sp1.test.es.elasticsearch.org/saml/acs",
  "attributes": {
     "department": [ "engineering", "product" ]
     "region": [ "APJ" ]
  }
}
```

There aren't any mapping explosion issues with that - we never store those requests as documents.

lloydmeta committed:
Simplify SAML attributes representation using JSON object/Map structure

Also, replace internal Attribute class list with a simpler Map<String, List<String>>
structure

This change:

- Removes the redundant Attribute class and replaces it with a direct Map
  implementation for storing attribute key-value pairs
- Eliminates the duplicate "attributes" nesting in the JSON structure
- Simplifies attribute validation without needing duplicate key checking

- Updates all related tests and integration points to work with the new structure

Before:

```js
{
  // others
  "attributes": {
    "attributes": [
      {
        "key": "department",
        "values": ["engineering", "product"]
      }
    ]
  }
}

After:

```js
{
  // other
  "attributes": {
    "department": ["engineering", "product"]
  }
}
```

(Verified by spitting out JSON entity in IdentityProviderAuthenticationIT.generateSamlResponseWithAttributes
... saw `{"entity_id":"ec:123456:abcdefg","acs":"https://sp1.test.es.elasticsearch.org/saml/acs","attributes":{"department":["engineering","product"],"region":["APJ"]}}`)

Signed-off-by: lloydmeta <lloydmeta@gmail.com>

lloydmeta commented on x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java:
```diff
@@ -199,6 +230,16 @@ private AttributeStatement buildAttributes(UserServiceAuthentication user) {
             attributes.add(name);
         }
+        // Add custom attributes if provided
+        if (customAttributes != null && customAttributes.getAttributes().isEmpty() == false) {
```
This will fail the linter because it wants explicit bool checks

lloydmeta committed:
* Fix up toString dangling quote.

Signed-off-by: lloydmeta <lloydmeta@gmail.com>

lloydmeta committed:
Merge branch 'main' into add-arbitrary-attribute-support

tvernum commented on x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java:
```diff
@@ -199,6 +230,16 @@ private AttributeStatement buildAttributes(UserServiceAuthentication user) {
+        // Add custom attributes if provided
+        if (customAttributes != null && customAttributes.getAttributes().isEmpty() == false) {
+            for (Map.Entry<String, List<String>> entry : customAttributes.getAttributes().entrySet()) {
+                Attribute attribute = buildAttribute(entry.getKey(), entry.getKey(), entry.getValue());
```
I think we should skip the `friendlyName` rather than set it to the same value as the formal name.

tvernum commented on x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/support/SamlInitiateSingleSignOnAttributes.java:
```diff
@@ -0,0 +1,146 @@
+    private Map<String, List<String>> attributes;
+
+    public SamlInitiateSingleSignOnAttributes() {
+        this.attributes = new HashMap<>();
```
I'd prefer `attributes` was final and immutable, and fully initialised in the constructor.

lloydmeta replied:
Done.

tvernum commented on x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/support/SamlInitiateSingleSignOnAttributes.java:
```diff
@@ -0,0 +1,146 @@
+     * @return A new SamlInitiateSingleSignOnAttributes instance
+     */
+    public static SamlInitiateSingleSignOnAttributes fromXContent(XContentParser parser) throws IOException {
+        Map<String, List<String>> attributes = new HashMap<>();
+        while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+            String key = parser.currentName();
+            if (parser.nextToken() == XContentParser.Token.START_ARRAY) {
+                List<String> values = new ArrayList<>();
+                while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                    values.add(parser.text());
+                }
+                attributes.put(key, values);
+            }
+        }
+        SamlInitiateSingleSignOnAttributes attributesObj = new SamlInitiateSingleSignOnAttributes();
+        attributesObj.setAttributes(attributes);
+        return attributesObj;
```
You should be able to replace all of this with:
```
        final Map<String, List<String>> attributes = parser.map(HashMap::new, p -> XContentParserUtils.parseList(p, XContentParser::text));
        return new SamlInitiateSingleSignOnAttributes(attributes);
```

lloydmeta replied:
Done thanks.

lloydmeta committed:
Merge branch 'main' of github.com:elastic/elasticsearch into add-arbitrary-attribute-support

lloydmeta committed:
* Remove attributes from Response object.

Signed-off-by: lloydmeta <lloydmeta@gmail.com>

lloydmeta committed:
* Remove friendly name.
* Make attributes map final in SamlInitiateSingleSignOnAttributes

Signed-off-by: lloydmeta <lloydmeta@gmail.com>

lloydmeta committed:
* Cleanup serdes by using existing utils in the ES codebase

Signed-off-by: lloydmeta <lloydmeta@gmail.com>

lloydmeta commented on x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/action/SamlInitiateSingleSignOnResponse.java:
```diff
@@ -20,19 +21,32 @@ public class SamlInitiateSingleSignOnResponse extends ActionResponse {
     private final String entityId;
     private final String samlStatus;
     private final String error;
+    private final SamlInitiateSingleSignOnAttributes attributes;
```
No good one. Removed.

lloydmeta commented on x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/support/SamlInitiateSingleSignOnAttributes.java:
```diff
@@ -0,0 +1,146 @@
+            }
+            builder.endArray();
+        }
+        return builder;
```
Done. Not sure how much it matters but I had to re-jig tue UTs to not start and end objects (e.g.         `builder.startObject();`) to make this change. I also noticed a `mapContents` method that looked like it could work without changing the tests but went with what you recommended.

lloydmeta commented on x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/support/SamlInitiateSingleSignOnAttributes.java:
```diff
@@ -0,0 +1,146 @@
+                values.add(in.readString());
+            }
+            attributes.put(key, values);
+        }
```
Done.

lloydmeta committed:
Touchup comment

Signed-off-by: lloydmeta <lloydmeta@gmail.com>

tvernum reviewed (commented):
LGTM, I'll try to get the transport versioning done tomorrow so we can merge.

lloydmeta committed:
Update x-pack/plugin/identity-provider/src/test/java/org/elasticsearch/xpack/idp/action/SamlInitiateSingleSignOnRequestTests.java

Co-authored-by: Tim Vernum <tim@adjective.org>

lloydmeta committed:
Merge branch 'main' into add-arbitrary-attribute-support

tvernum committed:
Merge branch 'main' into add-arbitrary-attribute-support

tvernum committed:
Add transport-version checks

tvernum reviewed (approved):
LGTM

ywangd reviewed (approved):
LGTM

I reviewed mostly the transport version changes and skimmed through other production code changes. Didn't review tests. I believe those are reviewed by Tim and others.
